stdlib/javascript/compile.mc

```diff
@@ -43,73 +46,159 @@ let _charSeq2String = use MExprAst in lam tms.
+  sem compileCOp (opts: CompileJSOptions) (args: [JSExpr]) =
   -- Binary operators
-  | CAddi _
-  | CAddf _ -> JSEBinOp { op = JSOAdd {}, lhs = head args, rhs = last args }
-  | CSubi _
-  | CSubf _ -> JSEBinOp { op = JSOSub {}, lhs = head args, rhs = last args }
-  | CMuli _
-  | CMulf _ -> JSEBinOp { op = JSOMul {}, lhs = head args, rhs = last args }
-  | CDivf _ -> JSEBinOp { op = JSODiv {}, lhs = head args, rhs = last args }
-  | CEqi _
-  | CEqf _  -> JSEBinOp { op = JSOEq {},  lhs = head args, rhs = last args }
-  | CLti _
-  | CLtf _  -> JSEBinOp { op = JSOLt {},  lhs = head args, rhs = last args }
-  | CGti _
-  | CGtf _  -> JSEBinOp { op = JSOGt {},  lhs = head args, rhs = last args }
-  | CLeqi _
-  | CLeqf _ -> JSEBinOp { op = JSOLe {},  lhs = head args, rhs = last args }
-  | CGeqi _
-  | CGeqf _ -> JSEBinOp { op = JSOGe {},  lhs = head args, rhs = last args }
-  | CNeqi _
-  | CNeqf _ -> JSEBinOp { op = JSONeq {}, lhs = head args, rhs = last args }
+  | CAddi _ & t
+  | CAddf _ & t -> optimizedOpIntrinsicGen t "add" args (_binOp (JSOAdd {}))
+  | CSubi _ & t
+  | CSubf _ & t -> optimizedOpIntrinsicGen t "sub" args (_binOp (JSOSub {}))
+  | CMuli _ & t
+  | CMulf _ & t -> optimizedOpIntrinsicGen t "mul" args (_binOp (JSOMul {}))
+  | CDivi _ & t
+  | CDivf _ & t -> optimizedOpIntrinsicGen t "div" args (_binOp (JSODiv {}))
+  | CModi _ & t -> optimizedOpIntrinsicGen t "mod" args (_binOp (JSOMod {}))
+  | CEqi  _ & t
+  | CEqf  _ & t -> optimizedOpIntrinsicGen t "eq" args (_binOp (JSOEq {}))
+  | CLti  _ & t
+  | CLtf  _ & t -> optimizedOpIntrinsicGen t "lt" args (_binOp (JSOLt {}))
+  | CGti  _ & t
+  | CGtf  _ & t -> optimizedOpIntrinsicGen t "gt" args (_binOp (JSOGt {}))
+  | CLeqi _ & t
+  | CLeqf _ & t -> optimizedOpIntrinsicGen t "le" args (_binOp (JSOLe {}))
+  | CGeqi _ & t
+  | CGeqf _ & t -> optimizedOpIntrinsicGen t "ge" args (_binOp (JSOGe {}))
+  | CNeqi _ & t
+  | CNeqf _ & t -> optimizedOpIntrinsicGen t "ne" args (_binOp (JSONeq {}))
 
   -- Unary operators
-  | CNegf _
-  | CNegi _ -> JSEUnOp { op = JSONeg {}, rhs = head args }
+  | CNegf _ & t
+  | CNegi _ & t -> optimizedOpIntrinsicGen t "neg" args (_unOp (JSONeg {}))
+
+  -- Sequential operators (SeqOpAst)
+  | CConcat _ -> intrinsicGen "concat" args
+  | CCons _ -> intrinsicGen "cons" args
+  | CFoldl _ -> intrinsicGen "foldl" args
+
+  -- Convert operations
+  | CChar2Int _ -> intrinsicGen "char2int" args
+  | CInt2Char _ -> intrinsicGen "int2char" args
+
+  -- References
+  | CRef _ -> intrinsicGen "ref" args
+  | CModRef _ -> intrinsicGen "modref" args
+  | CDeRef _ -> intrinsicGen "deref" args
 
   -- Not directly mapped to JavaScript operators
   | CPrint _ ->
-    JSEApp { fun = _consoleLog, args = args }
+    match opts.targetPlatform with CompileJSTP_Node () then intrinsicNode "print" args
+    else
+      -- Warning about inconsistent behaviour
+      printLn "Warning: CPrint might have unexpected behaviour when targeting the web or a generic JS runtime";
```

Comment:
You could provide a better warning message by passing the `info` field of the `TmConst` to the `compileCOp` (which should be called `compileJSOp`, right?). Then you could use the `info2str` function from `stdlib/mexpr/info.mc` to show where in the file the CPrint was used.